Shared: Add HMAC and cert utils

roosterfish opened the pull request:
This PR adds additional HMAC and cert utils to the `shared` package.

The HMAC implementation follows the practices from the [explicit trust establishment spec](https://discourse.ubuntu.com/t/explicit-trust-establishment-mechanism-for-microcloud/44261).
It's imported as a general utility in LXD to be used by MicroCloud during trust establishment.

In order to make the use of a KDF (in this case argon2) optional, there is a second factory function that allows using some extended functionality added by argon2.

To create a HMAC header, the caller runs:

```go
// Specific HMAC version to allow indicating protocol changes in the future.
const HMACMicroCloud10 trust.HMACVersion = "MicroCloud-1.0"

// Basic format
hmac := trust.NewHMAC(<key>, NewDefaultHMACConf("LXD1.0"))
// Create the "Authorization" header in the format `Authorization: <version> <HMAC>`
header, err := trust.HMACAuthorizationHeader(hmac, <message body to be sent>)

// Or custom format that extends HMAC with a KDF (in this case argon2).
hmac, err := trust.NewHMACArgon2(<password>, <salt or nil>, NewDefaultHMACConf("LXD1.0"))
// Create the "Authorization" header in the format `Authorization: <version> <salt>:<HMAC>`
header, err := trust.HMACAuthorizationHeader(hmac, <message body to be sent>)
```

On the server side the validation happens as follows:

```go
// Initialize the expected HMAC format
hmac, err := trust.NewHMACArgon2(<password>, <salt or nil>, NewDefaultHMACConf("LXD1.0"))

// Validate the HMAC based on the given *http.Request r
err = trust.HMACEqual(hmac, r)
```

roosterfish commented:
I am still not sure if we want to track the version within the format. Thinking about possible scenarios here.

tomponline commented on shared/cert.go:
```diff
@@ -430,19 +430,24 @@ func GenerateMemCert(client bool, options CertOptions) ([]byte, []byte, error) {
 }
 
+// ParseCert parse a X.509 certificate from the given byte slice and return its parsed content.
+func ParseCert(cert []byte) (*x509.Certificate, error) {
```
Would it be useful to adapt this to take an io.Reader instead so that a bytes.Buffer or actual file handle can be used? e.g. https://pkg.go.dev/bytes#example-Buffer-Reader

roosterfish replied:
It looks that the upstream `encoding/pem` is a limitation here. Whilst it accepts a writer for encoding, decoding is only supported using a `[]byte` slice. See https://pkg.go.dev/encoding/pem#Decode. 

So even if we accept an `io.Reader` here, we would still need to read it's contents and create a slice before we can decode the contents.

tomponline reviewed (changes requested):
Please could we have some tests that are usage examples for this package?

tomponline commented:
> This PR adds additional HMAC and cert utils to the `shared` package.
> 
> The HMAC implementation follows the practices from the [explicit trust establishment spec](https://discourse.ubuntu.com/t/explicit-trust-establishment-mechanism-for-microcloud/44261). It's imported as a general utility in LXD to be used by MicroCloud during trust establishment.
> 
> In order to make the use of a KDF (in this case argon2) optional, a format interface is added to allow plugging this functionality in on demand.
> 
> To create a HMAC header, the caller will run:
> 
> ```go
> // Specific HMAC version to allow indicating protocol changes in the future.
> const HMACMicroCloud10 trust.HMACVersion = "MicroCloud-1.0"
> 
> // Custom format that extends HMAC with a KDF (in this case argon2).
> argonFormat, err := trust.NewDefaultHMACFormatArgon()
> if err != nil {
>     return err
> }
> 
> hmac := trust.NewHMAC(trust.NewDefaultHMACConf(<passphrase>, HMACMicroCloud10)).WithFormat(argonFormat)
> 
> // Create the "Authorization" header in the format `Authorization: <version> <salt>:<HMAC>`
> header, err := hmac.AuthorizationHeader(<message body to be sent>)
> if err != nil {
>     return err
> }
> ```
> 
> On the server side the validation happens as follows:
> 
> ```go
> argonFormat, err := trust.NewDefaultHMACFormatArgon()
> if err != nil {
>     return response.InternalError(fmt.Errorf("Failed to setup argon: %w", err))
> }
> 
> requestHMAC := trust.NewHMAC(trust.NewDefaultHMACConf(<passphrase>, HMACMicroCloud10)).WithFormat(argonFormat)
> 
> // Validate the HMAC based on the given *http.Request r
> err = requestHMAC.ValidateAuthorizationHeader(r)
> if err != nil {
>     return response.SmartError(err)
> }
> ```

Is the expectation that this will be used in an http middleware? 

I'm a bit concerned how this is tightly coupled to the http package (although I do recall as discussing the balance of making it flexible and easy to test vs making it opinionated for the microcloud trust scenario).

Again, im wondering a bytes.Buffer would help to decouple this from the http package, and make setting/getting the Authorization header an excercise for the caller?

roosterfish commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,251 @@
+)
+
+// HMACVersion indicates the version used for the authorization header format.
+type HMACVersion string
```
Fixed.

roosterfish commented:
> Please could we have some tests that are usage examples for this package?

Added. I have split the tests into HMAC creation and validation to demonstrate both usage scenarios.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,265 @@
+type HMACFormatDefault struct{}
+
+// HMACFormatArgon represents the argon2 format.
+type HMACFormatArgon struct {
```
Shall we call this Argon2 rather than Argon to be clear what it implements?

roosterfish replied:
Fixed.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,265 @@
+
+// WithFormat configures the HMAC instance to use the specified format.
+// The format can be used to extend the way of displaying and parsing the resulting header.
+func (h *HMAC) WithFormat(format HMACFormat) *HMAC {
```
Can we just have different factory functions for the different formats, rather than having the ability to switch between them. I'm finding the method chaining functionality here a bit difficult to comprehend why it is needed?

roosterfish replied:
Fixed. There is now a `NewHMAC` and `NewHMACArgon2` which both return a struct that implements the `HMACFormatter`.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,265 @@
+// The Passphrase function is invoked so that the format can learn about the secret
+// used for generating the HMAC and can apply modifications such as running it through a KDF.
+// Example using argon2: `Authorization: <version> <salt>:<HMAC>`.
+type HMACFormat interface {
```
Lets call this HMACFormatter so it has a verb in it, which is ideally what interface names should have.

roosterfish replied:
Fixed.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,265 @@
+}
+
+// HMACFormatDefault represents the default format.
+type HMACFormatDefault struct{}
```
Please can you separate the two different hmac format options and their functions into separate files, e.g. hmac_default.go and hmac_argon2.go for clarity? You can keep the common entities in hmac.go as they are now.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,265 @@
+}
+
+// Header returns the actual HMAC alongside together with the used version.
+func (h *HMACFormatDefault) Header(version HMACVersion, hmac []byte) string {
```
Maybe call this HTTPHeader as to indicate where it should be used?

roosterfish replied:
Fixed.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,265 @@
+// NewDefaultHMACFormatArgon returns the default format for argon2.
+// Recommended defaults according to https://www.rfc-editor.org/rfc/rfc9106#section-4-6.2.
+// We use the second recommended option to not require a system having 2 GiB of memory.
+func NewDefaultHMACFormatArgon() (*HMACFormatArgon, error) {
```
Can this take a salt argument which can avoid generating a random one if provided?

roosterfish commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,265 @@
+	}
+
+	return HMACVersion(authHeaderSplit[0]), hmac, nil
+}
```
Fixed.
Empty strings are now forbidden for the version, HMAC and salt (in case of argon2).
I have also added a few more test cases for these scenarios.

Fedqys replied:
thanks !

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,205 @@
+	WriteJSON(v any) ([]byte, error)
+	WriteRequest(r *http.Request) ([]byte, error)
+
+	splitVersionFromHMAC(header string) (HMACVersion, string, error)
```
Does this need to be in the interface? Its not exported.

roosterfish replied:
No this doesn't need to be in the interface.
I put it there as I wanted to access it from all the implementations of the interface but realized it's enough to have it as part of the base struct as this one is already embedded in the argon struct.

Fixed and removed.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,205 @@
+	// HTTPHeader expects the HMAC computed over the payload and returns the final Authorization header.
+	HTTPHeader(hmac []byte) string
+	// ParseHeader expects an Authorization header and returns the used version and HMAC.
+	ParseHTTPHeader(header string) (HMACVersion, []byte, error)
```
comment/name mismatch

roosterfish replied:
Fixed.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,205 @@
+}
+
+// HMACAuthorizationHeader returns the HMAC on v using the given formatter.
+func HMACAuthorizationHeader(hmac HMACFormatter, v any) (string, error) {
```
the `hmac` argument covers the `hmac` package import - a small code smell.

roosterfish replied:
Fixed. Also in the `HMACEqual` func.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,205 @@
+	return nil
+}
+
+// HMACAuthorizationHeader returns the HMAC on v using the given formatter.
```
Returns the HMAC as an Authorization header?

roosterfish replied:
Fixed.

tomponline commented on shared/trust/hmac.go:
```diff
@@ -0,0 +1,205 @@
+
+	hmacFromBody, err := hmac.WriteRequest(r)
+	if err != nil {
+		return api.StatusErrorf(http.StatusInternalServerError, "Failed to write request body: %w", err)
```
This sounds a bit like its writing output in response to the request.

How about "Failed to calculate HMAC from request body"?

roosterfish replied:
Fixed.

roosterfish replied:
Thanks for the suggestion, I was thinking about an alternative here but couldn't find a good alternative.

tomponline reviewed (approved):
Thanks LGTM
